reference/src/clientimpl/src/sa_key_provision.c

```diff
@@ -0,0 +1,364 @@
+            INFO("cert length: %d", ((PlayReadyProvisioning*)(in))->modelCertificateLength);
+            INFO("in_length: %d", in_length);
+
+            if(NULL == ((PlayReadyProvisioning*)in)->privateKey             ||
+               0 == ((PlayReadyProvisioning*)in)->privateKeyLength          ||
+               NULL == ((PlayReadyProvisioning*)in)->modelCertificate       ||
+               0 == ((PlayReadyProvisioning*)in)->modelCertificateLength)   {
+               ERROR("null private key / cert or input size is  0");
+               return SA_STATUS_NULL_PARAMETER;
+            }
+
+            prProvision = calloc(sizeof(PlayReadyProvisioning), 1);
+            if(NULL == prProvision) {
+               ERROR("OOM");
+               return SA_STATUS_INTERNAL_ERROR;
+            }
+
+            /*deep copy*/
+            prProvision->privateKeyLength =
+               ((PlayReadyProvisioning*)in)->privateKeyLength;
+            prProvision->modelCertificateLength =
+               ((PlayReadyProvisioning*)in)->modelCertificateLength;
+
+            prProvision->privateKey =
+               (uint8_t*)calloc(prProvision->privateKeyLength, 1);
+            if(NULL == prProvision->privateKey) {
+               ERROR("OOM");
+               return SA_STATUS_INTERNAL_ERROR;
+            }
+            memcpy(prProvision->privateKey, ((PlayReadyProvisioning*)in)->privateKey,
+                   prProvision->privateKeyLength);
+
+            /*here just provide an examlpe to pass certification data
+              and model type,how to use them is all up to SOC vendors.
+            */
+            prProvision->modelCertificate =
+               (void*)calloc(prProvision->modelCertificateLength, 1);
+            if(NULL == prProvision->modelCertificate) {
+               ERROR("OOM");
+               return SA_STATUS_INTERNAL_ERROR;
+            }
+            memcpy(prProvision->modelCertificate,
+                   ((PlayReadyProvisioning*)in)->modelCertificate,
+                   prProvision->modelCertificateLength);
+
+            INFO("prProvision:%x",prProvision);
+            INFO("keyLen : %d", prProvision->privateKeyLength);
+            INFO("CertLen : %d", prProvision->modelCertificateLength);
+            INFO("Model type : %s",
+                  (prProvision->modelType == PLAYREADY_MODEL_2K)? "model 2000" :
+                  ((prProvision->modelType == PLAYREADY_MODEL_3K)? "model 3000":
+                   "No such model type"));
+
+            if(0 == (memcmp(prProvision->privateKey, "SIGN", 4))) {
+               prProvision->privateKey += 8;
+               prProvision->privateKeyLength -=8;
+            } else {
+              INFO("NO SIGN in the header");
+            }
+
+            sa_key *rsa_key = create_uninitialized_sa_key();
+            assert(NULL != rsa_key);
+            status = sa_key_import(rsa_key, SA_KEY_FORMAT_EXPORTED,
```

Comment:
Shouldn't the parameters field provide the SA_KEY_FORMAT_SOC type to denote the key is protected with the SOC key?